### ISSUE TITLE
ghc-options applies to all dependencies, not just local packages

### ISSUE THREAD

**Issue by edsko** [haskell/cabal#3883]
With traditional build, I would use

```
cabal build --ghc-option=-ddump-simpl
```

If I try the same with `new-build`, it starts to rebuild every single dependency; presumably because it considers the compiler flags for all dependencies to have changed. I tried to do

```
packages: ingest.cabal

profiling: True
executable-profiling: True
constraints: Cabal >= 1.24

program-options
  ghc-options: -fprof-auto 

package ingest
  ghc-options: -ddump-simpl -ddump-to-file
```

but although it doesn't complain about the `ghc-options` inside the `package` section, it doesn't seem to take it into account either; no dump files get created. What is the right approach? 

---- 

**edit by @23Skidoo:** [Currently recommended workaround](https://github.com/haskell/cabal/issues/3883#issuecomment-505394322) is to add a `cabal.project` that enables `ghc-options` for each package separately:

```
package foo
  ghc-options: -Werror

package bar
  ghc-options: -Werror
```

Cabal's own CI uses this approach.

**Comment by 23Skidoo** [comment]
My intuition is that the second option ought to work, so we should fix if it doesn't. Related: I've long wanted a `cabal ghc-core ModuleName` command.

**Comment by dcoutts** [comment]
The prog options should apply only to local packages, not all deps, just as all other command line or config top level options do.

**Comment by edsko** [comment]
Ah, turns out this is a duplicate of #3891 .

**Comment by dcoutts** [comment]
> With traditional build, I would use
>
>     cabal build --ghc-option=-ddump-simpl
>
>  If I try the same with new-build, it starts to rebuild every single dependency; presumably because it considers the compiler flags for all dependencies to have changed.

This is still a bug.

**Comment by ezyang** [comment]
See also #3720

**Comment by ezyang** [comment]
Bumping priority, this is pretty major and makes top-level ghc-options unusable.

**Comment by omefire** [comment]
Interesting one. I'd like to take a look and see if I can help fix this.

**Comment by ezyang** [comment]
Thanks @omefire, much appreciated! `programOptionsSectionDescr` at `cabal-install/Distribution/Client/ProjectConfig/Legacy.hs` is a good place to start (then just track how the config gets propagated through the code.)

**Comment by ezyang** [comment]
Hey @omefire, any progress / anything we can do to help?

**Comment by omefire** [comment]
@ezyang Sorry, I've been busy with moving to California and haven't had time to work on it. I'll get back to it by Friday evening or Saturday morning.

**Comment by omefire** [comment]
I wasn't able to reproduce this bug. Here's what I did:
```
- git clone https://github.com/jaspervdj/patat.git
- cd patat; cabal new-build
- modify patat.cabal by adding a new entry to the Ghc-options field
- rerun cabal new-build
```
Instead of rebuilding the package, I get a message `Up to date`.

**Comment by ezyang** [comment]
Well, the original bug report was referring to ghc-options in cabal.project. What if you put it there?

**Comment by omefire** [comment]
Ok, I wasn't aware of those two different files. can you point me to any documentation regarding cabal.project or project.cabal files ? and what the differences between them are ?

**Comment by ezyang** [comment]
Let me give you a counter offer: would you like a detailed description of what to do, similar to what I've written in #4047 and #4052?

(To answer your question, the docs are here: http://cabal.readthedocs.io/en/latest/nix-local-build.html#configuring-builds-with-cabal-project )

**Comment by omefire** [comment]
Thanks for the doc, I'll try working at it first and if I'm blocked, I'll request a detailed description.

**Comment by omefire** [comment]
I have been able to use `cabal new-build` and seen that some of the external deps (any package except for `cabal-install`) do get rebuilt when I include the `ghc-options` flag changes.
However, I just wanna make sure I'm going down the right path.

This is the behavior I get:
```
1- $ ~/Projects/cabal $ cabal new-build cabal-install
Resolving dependencies...
In order, the following will be built (use -v for more details):
Cabal-1.25.0.0
hackage-security-0.5.2.2
cabal-install-1.25.0.0
Configuring Cabal-1.25.0.0...
Preprocessing library Cabal-1.25.0.0...
[  1 of 145] Compiling Distribution.Utils.UnionFind ( Distribution/Utils/UnionFind.hs, /Users/omefire/Projects/cabal/dist-newstyle/build/Cabal-1.25.0.0/build/Distribution/Utils/UnionFind.o )
...

2- cabal new-build cabal-install
...
Up to date

3- $ cabal new-build cabal-install --ghc-options=-fno-ignore-asserts
Resolving dependencies...
In order, the following will be built (use -v for more details):
hackage-security-0.5.2.2
cabal-install-1.25.0.0
Configuring hackage-security-0.5.2.2...
Preprocessing library hackage-security-0.5.2.2...
Creating package registration file:
/Users/.../cabal/dist-newstyle/tmp/package-registration-1148079871137522503
Preprocessing executable 'cabal' for cabal-install-1.25.0.0...
Generating manual page
/Users/.../cabal/dist-newstyle/build/cabal-install-1.25.0.0/build/cabal/cabal.1

```

**Comment by ezyang** [comment]
> I haven't been able to use cabal build with most packages successfully due to some issues with cabal not building due to some deps conflicts (base or ghc). so, I haven't been able to verify for myself if this behavior appears when using cabal build.

Uh, that shouldn't happen. What were you doing that caused this error?

> This is the behavior I get:

So, I'm not entirely sure I can interpret your output, because I don't know what is pre-existing in your store, and consequently, I can't tell if something is or isn't being built depending on whether or not it is already installed. And in any case, the Cabal repository is not a great project to try doing reproduction on because it takes so long to build, and has many dependencies.

So here is the testing plan I would advise:

1. Create a small cabal.project with two packages, one which depends on the other
2. Set your HOME to a temporary directory, so you don't get false results from packages already being in your global store.
3. Run `cabal new-build` on your project. Then, run `cabal new-build --ghc-options=-fno-ignore-asserts` and verify that Cabal plans to rebuild both packages. (Due to #3891, GHC may not actually rebuild the packages. That's fine: the main question is whether or not Cabal is invoking GHC; anything else is GHC's problem.)
4. Reset your build environment (blow away the .store from your temporary HOME and dist-newstyle)
5. Run `cabal new-build` on your project. Now, modify `cabal.project` so that there is a `package` stanza that applies `ghc-options` to the downstream package (i.e., if p depends on q, apply the ghc-option to p). Run `cabal new-build` and verify that only p is rebuilt.
6. Reset your build environment again. Now, add a dependency to p on some quick to build, low dependency external package, THAT IS NOT in the system package store (or specify a version bound that excludes the system package from the solver.)  Twiddle ghc-options and check if the depended upon thing rebuilds. (It SHOULD NOT.)

Does that help?

**Comment by omefire** [comment]
@ezyang Thanks for the detailed and concise plan. It was very helpful.

I just went through your steps and the external dependency package in `step 6` doesn't get rebuilt.

I tested both with `cabal@1.24.0` and the current master as of commit `915ca233bd8bee36d3d8fb763e746455dfdfaae5`

**Comment by ezyang** [comment]
OK, so you've made a very good observation: the behavior *differs* depending on whether or not it was put in the cabal.project file versus on the command line (to double check, you did put ghc-options in the program-options stanza?) So the behavior in the command line case is a bug, I think!

**Comment by edsko** [comment]
This is not the behaviour I observe:

```
# cabal new-build 
Up to date
```

Change `cabal.project` and add

```
program-options
  ghc-options: -fprof-auto
```

Then

```
# cabal new-build --dry-run
Resolving dependencies...
In order, the following would be built (use -v for more details):
 - abstract-par-0.3.3 (lib) (requires build)
 - ansi-terminal-0.6.2.3 (lib) (requires build)
 - appar-0.1.4 (lib) (requires build)
 - auto-update-0.1.4 (lib) (requires build)
 - base-compat-0.9.1 (lib) (requires build)
...
```

(almost 180 dependencies would be rebuilt). This is with @hvr's `cabal-head-install`, 1.25+git20161109.0.d48c1eb-3~16.04 .

**Comment by omefire** [comment]
I stand corrected. I observe the same behavior as @edsko. I was using the wring stanza. So, both the cmd line and program-options yield the same behavior: rebuilding of external deps.

**Comment by phadej** [comment]
See also https://github.com/haskell/cabal/issues/3891#issuecomment-254408131

**Comment by omefire** [comment]
Just some updates on this issue:
I'm still working on it whenever I get some free time. 
I've been following/reading through the code from the `Main` module to `Cmd.Build` and others.
I'm currently looking into how the solver works in order to find out the root cause why the config option triggers a rebuild of external deps.

**Comment by ezyang** [comment]
OK, after some reading, I think the buggy code is `phaseConfigureCompiler` in `Distribution/Client/ProjectPlanning.hs`. Look at the invocation of it in `rebuildInstallPlan`:

```
          (projectConfig, projectConfigTransient) <- phaseReadProjectConfig
          localPackages <- phaseReadLocalPackages projectConfig
          compilerEtc   <- phaseConfigureCompiler projectConfig
          _             <- phaseConfigurePrograms projectConfig compilerEtc
          (solverPlan, pkgConfigDB)
                        <- phaseRunSolver         projectConfigTransient
                                                  compilerEtc
                                                  localPackages
          (elaboratedPlan,
           elaboratedShared) <- phaseElaboratePlan projectConfigTransient
                                                   compilerEtc pkgConfigDB
                                                   solverPlan
                                                   localPackages
```

The compiler is computed once, and then passed onto phaseElaboratePlan to be used as the compiler configuration information for all builds. Among other things, `compilerEtc` contains the `ProgramDb`, which records custom extra options that we pass along when we say `--ghc-options`, etc.

But if we look at `phaseConfigureCompiler`, it is using information from `projectConfigLocalPackages` (the "local-packages only config") to configure the compiler:

```
    phaseConfigureCompiler :: ProjectConfig
                           -> Rebuild (Compiler, Platform, ProgramDb)
    phaseConfigureCompiler ProjectConfig {
                             projectConfigShared = ProjectConfigShared {
                               projectConfigHcFlavor,
                               projectConfigHcPath,
                               projectConfigHcPkg
                             },
                             projectConfigLocalPackages = PackageConfig {
                               packageConfigProgramPaths,
                               packageConfigProgramArgs,
                               packageConfigProgramPathExtra
                             }
                           } = do
        progsearchpath <- liftIO $ getSystemSearchPath
        rerunIfChanged verbosity fileMonitorCompiler
                       (hcFlavor, hcPath, hcPkg, progsearchpath,
                        packageConfigProgramPaths,
                        packageConfigProgramArgs,
                        packageConfigProgramPathExtra) $ do

          liftIO $ info verbosity "Compiler settings changed, reconfiguring..."
          result@(_, _, progdb') <- liftIO $
            Cabal.configCompilerEx
              hcFlavor hcPath hcPkg
              progdb verbosity

        -- Note that we added the user-supplied program locations and args
        -- for /all/ programs, not just those for the compiler prog and
        -- compiler-related utils. In principle we don't know which programs
        -- the compiler will configure (and it does vary between compilers).
        -- We do know however that the compiler will only configure the
        -- programs it cares about, and those are the ones we monitor here.
          monitorFiles (programsMonitorFiles progdb')

          return result
      where
        hcFlavor = flagToMaybe projectConfigHcFlavor
        hcPath   = flagToMaybe projectConfigHcPath
        hcPkg    = flagToMaybe projectConfigHcPkg
        progdb   =
            userSpecifyPaths (Map.toList (getMapLast packageConfigProgramPaths))
          . userSpecifyArgss (Map.toList (getMapMappend packageConfigProgramArgs))
          . modifyProgramSearchPath
              (++ [ ProgramSearchPathDir dir
                  | dir <- fromNubList packageConfigProgramPathExtra ])
          $ defaultProgramDb
```

`packageConfigProgramArgs` is pulled out from `projectConfigLocalPackages`, fed into `progdb`, and then passed to `configCompilerEx` to actually configure the compiler. This is the correct thing to do for the compiler version (we want that to apply to ALL packages, including non-local ones), but it is totally not what we want for `ghc-options`, where now the options are being applied to non-local packages.

Fortunately, there is already existing code in new-build which tells us what to do here, namely the support for `ghc-options` in `package` stanza (although @edsko reports that `ghc-options` here doesn't work, that is actually an artifact of #3891; for example, our CI has very happily been using such options to enable `-Werror`). The answer is: PUNT. Program locations are controlled specifically by `with-ghc-location` (note the `-location` suffix) and if someone decides to change the GHC location for a specific package, and its a different version, they get to keep all the pieces.

So, here is my suggested plan of action:

1. Make equivalents of `packageConfigProgramPaths`, `packageConfigProgramArgs` and `packageConfigProgramPathExtra` in `ProjectConfigShared`, and rewrite `phaseConfigureCompiler` to use those, removing its dependence on "local" config.  As the current implementation uses them to influence compilation of everything, they really do belong in the shared configuration. In effect, this simply formalizes the idea that `program-options` is applied globally. That's fine but maybe we should rename the stanza to `global-program-options` or something.
2. So we need to add back the functionality of a locally applied only program option. Actually, the code change to achieve this is very simple: in `packageSpecificOptionsSectionDescr` we see how `ghc-options` inside is supported inside packages: `programOptionsFieldDescrs` is used to generate the field descriptors we need. So just move that code *into* `legacyPackageConfigFieldDescrs` so that it is also picked up by its top-level use in `legacyProjectConfigFieldDescrs`. Finally, we have to make it so that `packageConfigProgramArgs` is actually applied to the package: look at the assignments of `elabProgramArgs` and `elabProgramPaths` in `ProjectPlanning` and you'll see `perPkgOptionMapLast` is used to pull out the per-package paths/args and mappend them into the global, compiler db. Just also mappend packageConfigProgramPaths and packageConfigProgramArgs from localPackagesConfig and that should be groovy
3. You need to write a test. Maybe wait for #4124 to merge.
3. Cherry on top: update our documentation to correctly document all of these program-options field. When I originally wrote the documentation I didn't understand how this code was setup. Once you make these changes, you will understand it well.

**Comment by ezyang** [comment]
@omefire Any progress? Need more help?

**Comment by phadej** [comment]
FWIW, there should be a way to specify `ghc-options` for building dependencies, and in the both flavours: causing rebuild if not match, and not.

**Comment by ezyang** [comment]
I opened a separate bug for the thing where when you say `--ghc-options=-ddump-simpl`, ASSUMING that the semantics of this is to apply globally, Cabal decides to rebuild everything.

**Comment by ttuegel** [comment]
I got bit by this recently, but before I take a crack at it, I want to make sure there is general agreement about what should be happening.

1. Command-line flags should apply only to the current targets. (Currently they apply to all packages.) This affects `ghc-options` and other flags because the command-line flags get shoved into the `ProjectBaseContext`: https://github.com/haskell/cabal/blob/master/cabal-install/Distribution/Client/ProjectOrchestration.hs#L162 I tried to work on this yesterday before I found this issue; this is actually pretty tricky to disentangle.
2. `cabal.project` and `cabal.project.local` should apply to packages in the current project. (This is currently the correct behavior.)
3. There should be a way to specify a persistent configuration that applies to all packages. For the original and sandbox commands, we have `cabal.config` in this role; perhaps that should apply to `new-*` commands also? (By the way, this would remove the need for #4646.)

**Comment by ttuegel** [comment]
> Command-line flags should apply only to the current targets.

Yeah, never mind, I'm not touching this hot mess until we have per-component build everywhere. ☹️

**Comment by vrom911** [comment]
Any updates on this one? This is very useful to have this issue implemented.

**Comment by harendra-kumar** [comment]
@23Skidoo It will be great if this issue can be prioritized. Because of this we cannot use `-Werror` option at all in CI builds, it tries to build everything with -Werror and many dependencies fail to build. I have been using `stack` builds with -Werror instead of cabal builds to avoid this issue.

**Comment by phadej** [comment]
I don't think this is high priority as there is a workaround by explicitly listing the packages

```
package foo
  ghc-options: -Werror

package bar
  ghc-options: -Werror
```

e.g. [`haskell-ci`](https://github.com/haskell-ci/haskell-ci) have a way to generate such "local package" options sections for CI's `cabal.project`.

---

Yet, it might be good idea to go through https://github.com/haskell/cabal/issues?q=is%3Aopen+is%3Aissue+label%3A%22priority%3A+high+%3Afire%3A%22 as there are not so much of those, and some seems to be outdated (by v2-build e.g.)

**Comment by phadej** [comment]
Note how `Cabal` own CI uses that workaround: https://github.com/haskell/cabal/blob/5a07388f299ddc55fea5ba61c1b5b74de473c9b9/cabal.project.travis

**Comment by chshersh** [comment]
@phadej But what if I don't want to introduce a separate file just to build my package on CI with `-Werror`? I personally don't need `cabal.project` that often. And, moreover, even if I have one, I probably need to introduce a separate file and duplicate all configurations because I don't want to build with `-Werror` locally.

**Comment by 23Skidoo** [comment]
IMO if you're using `-Werror` on CI, it makes sense to use it locally too. You can always disable it temporarily.

**Comment by chshersh** [comment]
I favour the approach where you use `-Wall` locally and `-Werror` on CI. During development, it's a bit tedious to get rid of all warnings every time: redundant imports, missing export lists, unused functions/types/fields/constructors, etc. But your final result shouldn't contain any warnings. Unfortunately, it's quite easy to miss them locally. Fortunately, this is a purpose of CI — catch all such things.

**Comment by harendra-kumar** [comment]
@phadej thanks for pointing out the workaround. I was not aware of it. This is useful.

**Comment by lexi-lambda** [comment]
A problem with the current workaround is that it’s impossible to list out all local packages in the global `~/.cabal/config` file. I’d like to enable `ghc-options: -j` for all local packages globally, but I don’t think that’s currently possible.

**Comment by cartazio** [comment]
What are the architectural or man power blockers here ?

**Comment by emilypi** [comment]
@lexi-lambda agreed - this would be great for tooling as well, since certain tools like [hsinspect](http://hackage.haskell.org/package/hsinspect) require plugins to be enabled in order to gather the correct data. Obviously committing build-tool specific details to source is a bad idea, but without being able to support flags and options with some global config or `cabal.project.local`-like behavior, I'm left in a suboptimal position. 

But, just so it doesn't like I'm complaining at developers, I'm available to discuss this more in depth and contribute. Just point me in the right direction 😄

**Comment by sol** [comment]
> Bumping priority, this is pretty major and makes top-level ghc-options unusable.
@ezyang agree!

> I don't think this is high priority as there is a workaround by explicitly listing the packages
@phadej disagree!

@hvr & co, instead of pointing fingers at others, lobbying against other peoples projects and hampering competition, please put your effort in producing working software yourself!

**Comment by cartazio** [comment]
Or more broadly: what contributor support or user feedback is needed to
help correctly triaging and assessing priorities?

**Comment by jneira** [comment]
After #7973, `ghc-options` is only applied to all local packages. It is still possible recover the old behaviour, see the workaround here: https://github.com/haskell/cabal/pull/7973#issuecomment-1038837581
Closing the issue, thanks all!
